### PR TITLE
[SPARK-38953][PYTHON][DOC] Document PySpark common exceptions / errors

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Document PySpark(SQL, pandas API on Spark, and Py4J) common exceptions/errors and respective solutions.

### Why are the changes needed?
Make PySpark debugging easier.

There are common exceptions/errors in PySpark SQL, pandas API on Spark, and Py4J.

Documenting exceptions and respective solutions help users debug PySpark.

### Does this PR introduce _any_ user-facing change?
No. Document change only.

### How was this patch tested?
Manual test.

image

Please refer to https://github.com/apache/spark/blob/7a1c7599a21cbbe2778707b72643cf98ac601ab1/python/docs/source/development/debugging.rst#common-exceptions--errors for the whole rendered page.

### CODE REVIEW

**Comment by xinrong-databricks** [comment]
Got an error 
```py
Warning, treated as error:
/__w/spark/spark/python/docs/source/development/debugging.rst:348:Could not lex literal_block as "python". Highlighting skipped.
make: *** [Makefile:35: html] Error 2
```
but the rendered file looks fine.

**Comment by xinrong-databricks** [comment]
@ueshin @HyukjinKwon @itholic Thank you!
Also CC @gengliangwang

File: python/docs/source/development/debugging.rst [hunk 1]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+AnalysisException is raised when failing to analyze a SQL query plan.
+
+Example:
+
+.. code-block:: python
+
+    >>> df = spark.range(1)
+    >>> df['bad_key']
+    Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:353]: Seems like we should change `…` to `...`.

File: python/docs/source/development/debugging.rst [hunk 2]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+
+ParseException is raised when failing to parse a SQL command.
+
+Example:
+
+.. code-block:: python
+
+    >>> spark.sql("select * 1")
+    Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:373]: ditto

File: python/docs/source/development/debugging.rst [hunk 3]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+
+IllegalArgumentException is raised when passing an illegal or inappropriate argument.
+
+Example:
+
+.. code-block:: python
+
+    >>> spark.range(1).sample(-1.0)
+    Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:397]: ditto

File: python/docs/source/development/debugging.rst [hunk 4]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+.. code-block:: python
+
+    >>> from pyspark.sql.functions import udf
+    >>> def f(x):
+    ...   return F.abs(x)
+    ...
+    >>> spark.range(-1, 1).withColumn("abs", udf(f)("id")).collect()
+    22/04/12 14:52:31 ERROR Executor: Exception in task 7.0 in stage 37.0 (TID 232)
+    org.apache.spark.api.python.PythonException: Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:424]: ditto

File: python/docs/source/development/debugging.rst [hunk 5]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+Example:
+
+.. code-block:: python
+
+    >>> sdf = spark.readStream.format("text").load("python/test_support/sql/streaming")
+    >>> from pyspark.sql.functions import col, udf
+    >>> bad_udf = udf(lambda x: 1 / 0)
+    >>> (sdf.select(bad_udf(col("value"))).writeStream.format("memory").queryName("q1").start()).processAllAvailable()
+    Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:450]: ditto

File: python/docs/source/development/debugging.rst [hunk 6]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+    org.apache.spark.api.python.PythonException: Traceback (most recent call last):
+      File "<stdin>", line 1, in <lambda>
+    ZeroDivisionError: division by zero
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:454]: ditto

File: python/docs/source/development/debugging.rst [hunk 7]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+Example:
+
+.. code-block:: python
+
+    >>> from pyspark.sql.functions import to_date, unix_timestamp, from_unixtime
+    >>> df = spark.createDataFrame([("2014-31-12",)], ["date_str"])
+    >>> df2 = df.select("date_str", to_date(from_unixtime(unix_timestamp("date_str", "yyyy-dd-aa"))))
+    >>> df2.collect()
+    Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:474]: ditto

File: python/docs/source/development/debugging.rst [hunk 8]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+
+Operations involving more than one series or dataframes raises a ValueError if “compute.ops_on_diff_frames” is disabled (disabled by default). Such operations may be expensive due to joining of underlying Spark frames. So users should be aware of the cost and enable that flag only when necessary.
+
+Exception:
+
+.. code-block:: python
+
+    >>> ps.Series([1, 2]) + ps.Series([3, 4])
+    Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:501]: ditto

File: python/docs/source/development/debugging.rst [hunk 9]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+
+.. code-block:: python
+
+    >>> def f(x) -> ps.Series[np.int32]:
+    ...   return x[:-1]
+    ...
+    >>> ps.DataFrame({"x":[1, 2], "y":[3, 4]}).transform(f)
+    22/04/12 13:46:39 ERROR Executor: Exception in task 2.0 in stage 16.0 (TID 88)
+    org.apache.spark.api.python.PythonException: Traceback (most recent call last):
+    …
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:528]: ditto

File: python/docs/source/development/debugging.rst (hunk 1 above)
Comment (xinrong-databricks) [reply python/docs/source/development/debugging.rst]: That's helpful! Thanks.

**Comment by itholic** [comment]
Would mind adding the screen-capture of the some part of page so that we can easily verify the page is rendered properly ??

**Comment by xinrong-databricks** [comment]
@itholic Good idea! Added.

File: python/docs/source/development/debugging.rst [hunk 10]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
+
+Common Exceptions / Errors
+--------------------------
+
+PySpark SQL
+~~~~~~~~~~~
+
+- **AnalysisException**
+
+AnalysisException is raised when failing to analyze a SQL query plan.
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:344]:
```suggestion
``AnalysisException`` is raised when failing to analyze a SQL query plan.
```

to highlight the name ?

File: python/docs/source/development/debugging.rst [hunk 11]
```diff
@@ -332,3 +332,273 @@ The UDF IDs can be seen in the query plan, for example, ``add1(...)#2L`` in ``Ar
...
 
 This feature is not supported with registered UDFs.
+
+Common Exceptions / Errors
+--------------------------
+
+PySpark SQL
+~~~~~~~~~~~
+
+- **AnalysisException**
```
Comment (itholic) [inline python/docs/source/development/debugging.rst:342]:
```suggestion
**AnalysisException**
```

Not strong feeling about this, but seems like the indentation is a bit unmatched with the contents below ??

File: python/docs/source/development/debugging.rst (hunk 10 above)
Comment (xinrong-databricks) [reply python/docs/source/development/debugging.rst]: Sounds good!

File: python/docs/source/development/debugging.rst (hunk 11 above)
Comment (xinrong-databricks) [reply python/docs/source/development/debugging.rst]: Good idea! Indentation is adjusted.

**Review by HyukjinKwon — COMMENTED** [review]
We might need to describe the configurations like:
- `spark.sql.execution.pyspark.udf.simplifiedTraceback.enabled`
- `spark.sql.pyspark.jvmStacktrace.enabled`
- log4j configuration to make it verbose or silent

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.3.

**Pull request closed without merging**